### ISSUE TITLE
Cosmos sdk throws 408 cosmosException, but operation succeded in the database

### ISSUE THREAD

**Issue by BjornDeRijcke** [Azure/azure-cosmos-dotnet-v3#1420]
**Description**
Using the v4 preview.

When requests take a long time to return a response, the sdk throws an exception with code 408 - Request Timeout. However, if you'd then go looking in the db, the operation actually happened.

Currently our case is we insert a new item in the db. Sometimes we get the 408 exception, but if we retry we get a 409 - Conflict response from the database (saying an entity with this id already exists). This because the operation actually succeeded on the database, the sdk just threw a timeout. 

So if you set a break point when such an exception is thrown. And then when it is hit, you go look in the database, you will see the document exists (and thus was created by the request that threw the 408 exception)

**To Reproduce**
We have the following code:
```
var policyResult = await _retryOnHeavyLoadPolicy.ExecuteAsync(async () =>
{
    await store.CreateItemAsync(entity, new PartitionKey(entity.Discriminator), null, cancellationToken);

    return Task.CompletedTask;
}).ConfigureAwait(false);
```
Where the heavy load retry policy is a Polly.Net policy defined as:
```
Policy
    .Handle<Exception>(x =>
        x is CosmosException
        && (((CosmosException)x).Status == (int)HttpStatusCode.TooManyRequests
            || ((CosmosException)x).Status == (int)HttpStatusCode.RequestTimeout))
    .WaitAndRetryForeverAsync(retryAttempt => TimeSpan.FromSeconds(Math.Pow(retryAttempt, 0.5)));
```
This policy will then retry the create statement if we gat 429 - TooManyRequest or a 408 - Timeout exception.

To **reproduce** you will need a pretty **heavy load and a bad / slow internet** connection though.

**Environment summary**
SDK Version: v4.0.0-preview3
OS Version: Windows 10

**Question**
Due to our retry policy, when creating an item we sometimes get a 408 -> retry -> 409. How should we handle the 408 exceptions that we are sure the item is really created in the database (or updated, deleted, ... any action)? Without retrying, because that causes a 409.

**Comment by BjornDeRijcke** [comment]
How should this be solved in our code? Can we be sure such a scenario will not happen when everything is running in the azure cloud? (no requests taking longer than the default timeout) ?

**Comment by bartelink** [comment]
you're better off working on the basis of https://en.wikipedia.org/wiki/Idempotence at all times - this makes you ready for the inevitable myriad ways in that you'll arrive at that need anyway. The other good thing about doing things correctly is that you'll correctly make your requests contingent on etags etc., which will. ehance your stability under load as retries return 304s vs triggering write costs at the worst possible time

**Comment by BjornDeRijcke** [comment]
I don't have the feeling Idempotence has much implications about this question. All that we are seeing here is that **the sdk signals us** that **the request didn't happen**. That's fine for us, we'll just try again. That's why we retry when we get a 429 or a 408. (the sdk already retries 429 under the hood anyway, so I don't see why you could consider this wrong). Using etags on an insert does not seem to make a lot of sense to me, since the entity is new en not yet created in the database - so there is no etag. Perhaps I understand you wrong on that account?

However the issue is that the sdk is 'lying' in a sense because the entity did end up in the database. But we still got the 408 - timeout. Ofcourse, this makes sense since http requests simply can take a long time to go back and forth - or get lost and need to be retransmitted etc. which means they obviously can take longer to complete that the set timeout.

If the entire domain logic would run, of course we'd know that the entity is already in the database if we needed to know it for validation. But that has nothing to do with the cosmos db sdk, so I'm not asking anything about that here.

What we want is at our **infrastructure level code** (= our code around cosmos sdk) to **retry the request if we have reason to believe we can**. With a 429 you definitely can retry the request and that's the suggested way. Perhaps with a 408 the request should simply be let to fail and the entire domain logic should be rerun if you want to retry the request. That's the whole point of me asking this question. I'd like the cosmos team's suggestion of how they see this scenario should be handled by users of their sdk

**Comment by bartelink** [comment]
While I dont completely disagree, I will calrify... 
- 408 never means anything didnt/wont happen in my book - if you ask for something, it might have happened.
- yes, I dont specifically mean etags for the insert case of course.

Regardless of whether an upstream told me 408, or 500, I would not expect a guarantee of nothing having completed - it's just not guaranteed (whereas I'd be surprised if a 429 actually did the work).

The only issue for me would be if the SDK was pretending a 408 came back from the Cosmos server when really an internal Timeout was exceeded -- that'd be a bug (lying in your terms) for me, and something under its control.

-with a 408 you can retry the request, but, as you suggest, this might represent a case where you want to propagate the timeout (and potential retry or fallback policies) outside
-with a 429 you can retry the request _subject to backing off per the header, which is why that's encapsulated in the SDK_

**Comment by ealsur** [comment]
Can you paste a full stack trace of the exception when it happens? Also, on which types of operations are you seeing the 408?

408 often means Timeout, so it can be related to a networking issue https://docs.microsoft.com/en-us/azure/cosmos-db/troubleshoot-dot-net-sdk#request-timeouts

**Comment by BjornDeRijcke** [comment]
I get the following exception + stacktrace:

```
Response status code does not indicate success: 408 Substatus: 0 
Reason: (
    Message: Request timed out.ActivityId: 27ff3cf4-6c1a-4ac2-bae6-1803d0c7eca8, 
    Request URI: /apps/DocDbApp/services/DocDbServer24/partitions/a4cb4964-38c8-11e6-8106-8cdcd42c33be/replicas/1p/, 
    RequestStats: 
        RequestStartTime: 2020-04-23T10:18:59.2530643Z, 
        RequestEndTime: 2020-04-23T10:18:59.2530647Z, 
        Number of regions attempted:1, 
        SDK: Windows/10.0.18362 cosmos-netstandard-sdk/3.4.1).

  at Microsoft.Azure.Cosmos.ResponseMessage.EnsureSuccessStatusCode()
   at Azure.Cosmos.CoreExtensions.EnsureSuccessStatusCode(Response response)
   at Azure.Cosmos.CosmosResponseFactory.ToObjectInternal[T](Response response, CosmosSerializer jsonSerializer)
   at Azure.Cosmos.CosmosResponseFactory.<CreateItemResponseAsync>b__9_0[T](Response cosmosResponseMessage)
   at Azure.Cosmos.CosmosResponseFactory.<ProcessMessageAsync>d__17`1.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Common.Infrastructure.Cosmos.Extensions.CosmosContainerExtensions.<CreateItemAsync>d__8`1.MoveNext() in C:\Repos\ZeroFriction\ZFH.Backend\Shared\Common\Infrastructure\Cosmos\Extensions\CosmosContainerExtensions.cs:line 227
```

When you look in the response, you can see there are no headers or anything from a remote.

As for the operations, we only noticed this for the creation of new objects. Because that was causing the id already exists issues. (409)
We would believe though that this is not limited t a single operation. However we have not yet encountered issues.

Will read through the document you linked.

**Comment by bartelink** [comment]
@BjornDeRijcke as a matter fo interest, do you also see this using V3 ?

**Comment by Adelzu** [comment]
> @BjornDeRijcke as a matter fo interest, do you also see this using V3 ?

I have this in V3

**Comment by Adelzu** [comment]
Side effect to this when it happens, We write an object (Another object exists with the same Id, across multiple different partitions, [which is allowed by Cosmos](https://docs.microsoft.com/en-us/azure/cosmos-db/databases-containers-items#properties-of-an-item))
- When it fails with 408, we retry again to make sure it is written to the DB
- It fails again, this time the error is conflict, no big deal. Object is for sure written in the DB. We need the _etag thgouh, which we didn't get in both attempts.
- Read the object using the same Container object providing the Id and PartitionKey.
- We get a wrong object from a different partition with the same Id (we have the same object Id

[This is the file we are having the issue with ](https://github.com/leiratech/Event-Sourcing-Core/blob/master/src/EventSourcing/Extensions/CosmosContainersExtensions.cs). Lines 42 through 58 are an ugly hack to get around this.

**Comment by ealsur** [comment]
There is something that does not make sense. You cannot get the wrong object from a different partition.
An item uniqueness is defined by its id AND Partition Key value. 
In a distributed system, if you have multiple writers that can write the same document (id + PK) then there is always the possibility of network blips, transient connection issues, that could lead to one client landing the request first.
But if you read afterwards, you always read with Id + PK, so you cannot read an item from a different Partition, because regardless of which client wrote it, the Id + PK was the same for both, and you are passing that on the Read operation.

**Comment by Adelzu** [comment]
@ealsur Agreed, it is absolutely a strange thing to happen. Write fast enough then read, and this happens.

Notice that the write is correct, we get an immediate conflict on the 2nd attempt write, which means the first write was successful and the database already knows that this document exists and could find it. However the read yields wrong result.
It is expected to get null when the document is not synced across regions, or when consistency is not strict and using a different session. But here, everything is using the same Container Object, and not only i get a result, i get a wrong one.

Here is a debug screenshot that shows it.
![image](https://user-images.githubusercontent.com/14052842/83055887-261c1a00-a05d-11ea-9224-cfb89ad83b04.png)

Also here is an [MSDN forum thread](https://social.msdn.microsoft.com/Forums/en-US/66930893-4a97-4ab9-ab7a-bca2adb386f4/inserting-duplicate-id-across-multiple-partitions-results-in-wrong-reads-immediately-after-insert?forum=azurecosmosdb) where i was discussing the issue with one of your team members and all explanations provided by the product team do not make sense.

**Comment by ealsur** [comment]
@Adelzu First, if the document is not synced or you use an old session, you'd get a 404 with no content. 
Have you tried printing the values that the `item` variable passes to the method call? Could the `item` variable be concurrently being used on different threads? This can cause this type of behavior when you inspect variables.
If the ReadItemAsync method was called with an item id and a particular PK value, there simply is no way the service returns a different value, this directly maps to a REST API call with those values.
The SDK is not doing anything special and if this is actually the case, it would be some issue on the service side.

Keep in mind that a document can be updated and the id of the document changed. Could you be doing a Replace operation and changing the id value of a document?

**Comment by ealsur** [comment]
Another thing you can verify is the `_ts` of the document. That will point to the last write operation and should match to the time when the write that succeeded happened. If it does not, then it was either modified afterwards or is not the document you think it is.

**Comment by Adelzu** [comment]
@ealsur 
> if the document is not synced or you use an old session, you'd get a 404 with no content.

Agreed, this is the expected behavior.

> Have you tried printing the values that the `item` variable passes to the method call? Could the `item` variable be concurrently being used on different threads? This can cause this type of behavior when you inspect variables.

The main item is created and never modified across the system. The partitionKey is set only once.  Notice that this behavior is not always apparent, it happens like 70% of the time. 
Even though everything happens in parallel, i verified that each path of this particular issue is thread safe. I didn't try printing as the output would be a mess since this specific ExtensionMethod is called so many times. So if you have ideas, i'm all ears.

> If the ReadItemAsync method was called with an item id and a particular PK value, there simply is no way the service returns a different value, this directly maps to a REST API call with those values.
Could the database be returning wrong values?

A bug in the indexing services is a plausible explanation. However i don't know the internal architecture of the database so can't just say for sure.

> The SDK is not doing anything special and if this is actually the case, it would be some issue on the service side.

Most probably it is on the service side. You guys can reach to the correct team, I'm unable to.

> Keep in mind that a document can be updated and the id of the document changed. Could you be doing a Replace operation and changing the id value of a document?

Even if Ids can change, our problem here is with PartitionKey, and that is immutable.

>Another thing you can verify is the _ts of the document. That will point to the last write operation and should match to the time when the write that succeeded happened. If it does not, then it was either modified afterwards or is not the document you think it is.

Since operations happen in parallel, there is a good chance of multiple items with the same Id created at the same time. so _ts is not really a good measure (Add also network latency, so this can't be accurate).

**Comment by ealsur** [comment]
> The main item is created and never modified across the system. The partitionKey is set only once. Notice that this behavior is not always apparent, it happens like 70% of the time.
Even though everything happens in parallel, i verified that each path of this particular issue is thread safe. I didn't try printing as the output would be a mess since this specific ExtensionMethod is called so many times. So if you have ideas, i'm all ears.

Try assigning local variables instead of using `item.Id` and `item.PartitionKey` directly, the same with the end result comparison. Your code is already doing a `Console.Write, that is why I thought you could add to the same write, those values.

Basically:

```csharp
string partitionKey = item.PartitionKey;
string id = item.Id;
using (ResponseMessage responseMessage = await container.ReadItemStreamAsync(id, new PartitionKey(partitionKey)))
{
    // Deserialize and compare the response's id and PK
}
```

> Most probably it is on the service side. You guys can reach to the correct team, I'm unable to.
If this is the case, a support ticket needs to be raised.

**Comment by ealsur** [comment]
Closing due to inactivity.
